### ISSUE TITLE
No possibility to specify grace period in the Healthcheck for service startup

### ISSUE THREAD

**Issue by vshcherb** [docker/docker#26664]
This is follow up of these issues and changes:
https://github.com/docker/docker/issues/21142
https://github.com/docker/docker/pull/23218
https://github.com/docker/docker/issues/24396
https://github.com/docker/docker/pull/24856

So following the changes, in the original request 21142 there was a grace-period mentioned to let service startup longer. There was a workaround implemented using exit code 2 in 23218 pull-request. Later to simplify structure of the healthcheck the exit code 2 support was dropped. Now we end up in a situation that we can't have different timeout period for healthcheck during startup and during running state. 

Why this is critical: 
- In a swarm mode if a service reports unhealthy certain amount of time,  it will be killed and restarted.
- On the other hand if a service reports healthy while it is being unhealthy it could cause crashes of dependent services (they discover the main service is healthy and start loading).

There is another reason why this is a generic healthcheck problem. Let's put have an imaginary container that on startup loads 1TB from hard drive into the RAM, so we know that it could take up to 20 minutes initial load, though the proper healthcheck should be responding in 1 seconds and we would like to restart service after 30 seconds it's being unhealthy.

What's tried: 
- Obviously returning exit code 2 doesn't work 
- The solution mentioned in this comment https://github.com/docker/docker/issues/24396#issuecomment-233073059 definitely doesn't work, it hangs the whole docker service on the host machine. 

Solutions:
- I don't insist that exit-code 2 was a good solution, I would say better solution is to specify different number of retries count initial and after successful run. Let's say

```
HEALTHCHECK --interval=10s --startup-retries=60 --retries=3 --timeout=10s CMD /bin/bash -x /healthcheck.sh
```

That would allow service to startup for 10 minutes and be unhealthy for 30 seconds which is a valid case. Of course proper healthcheck to handle that situation is needed.

**Comment by thaJeztah** [comment]
ping @aluzzardi @crosbymichael @talex5

**Comment by yongtang** [comment]
Created a PR #26673 that adds `--initial-retries` for `HEALTHCHECK`. Please take a look.

**Comment by justincormack** [comment]
I think the argument around this (removing code 2) was that the healthcheck was not there to show readiness, and so another mechanism should be used to show readiness, and so services should return healthy when they are starting up normally even if they are not ready to serve (eg we discussed database recovery which can take a very long time). In that case we should not fix this by suggesting that services should return unhealthy when they are starting up normally, but by adding a way to signal readiness.

**Comment by vshcherb** [comment]
I'm afraid adding concept readiness could be mislead people. I'm not arguing that's bad or good approach, as of today we have a third state "container starting" which can not be controlled by the docker image (i.e. docker build concept). So in terms of user, "container starting" is displayed when healthcheck returns unhealthy which leads developer to use HEALTHCHECK to implement "proper" starting state using it. 
I'm afraid if we introduce a different concept like READINESS it could be more confusing and more difficult to implement, I would say extra parameter like --initial-retries or --initial-timeout or to HEALTHCHECK is easier in both ways. 
The possible problem that it is hard to estimate start-up period and it is highly dependent on end-user application. Though  I don't expect that feature will be used in 95% of docker images, but it could help people in extraordinary cases and  who wants to use docker everywhere.

**Comment by kakawait** [comment]
I have exactly same issue, today I really can't `service create` safely tomcat container. Time between tomcat starting and application is successfully deploy and up and running is no the same and I think is not always predictable. Thus we can't rely on *time-base* solution, we need a way (as same as previous `exit 2`) to share **READINESS** status with container manager.

Otherwise *docker swarm mode* is not useable except on small sample/demo

**Comment by vshcherb** [comment]
@kakawait Even if you know that tomcat is not started, it is a better practice to specify timeout for startup process as well. Having unpredictable startup period (it simply could be > 1 day if something goes wrong) could create other issues.

**Comment by kakawait** [comment]
@vshcherb You are true. My main problem is actually I'm doing database migrations using [Flyway](https://flywaydb.org/). Flyway migrations are running during deployment and tomcat will wait end of migrations before starting listening HTTP requests.

Thus depending of app version (if new version contains or not migration) deployment can take less than 1 minute to more than 1 hour.

Maybe I should split app and migration on two differents app/process. But from what I heard swarm does not allow you to launch task (!= service) on cluster.

**Comment by bvis** [comment]
Hi, any news about this issue?

**Comment by kakawait** [comment]
I know it's easy to point feature from concurrent but what I wanted to explain and what I expect from that request is something similar to what kubernetes offers by splitting *liveness* and *readiness* healthcheck as explain here https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

Please don't answer me to move on kubernetes is not subject.

**Comment by vpusher** [comment]
👍  I ran into the exact same issue. This is a major blocker for us to move on Swarm on production. Having containers that start in an **unpredictable** time (first operations, data migration, ...) cause unhealthy states and kills. Not sure if time based (or retries) solutions are appropriates.

**Comment by justincormack** [comment]
The `HEALTHCHECK` mechanism is not a readiness mechanism, and if you try to use it like that your container will be killed. As the original issue says, a fixed time is not suitable for eg database recovery, and your health check must return success during database recovery - you must explicitly check to see what is going on. Otherwise you are better off removing the healthcheck.

Readiness is very different, it is very unclear that it wants the same interface, and people eg want support to start other containers when another is ready, which is a completely different interface. Using events might for example be more appropriate.

I am not sure adding a grace period is a good idea, given that people only seem to want to use it to emulate readiness, but this is still likely to get their containers killed and cause unrecoverable database corruption in the situations that people appear to want to use it.

**Comment by kakawait** [comment]
@justincormack I'm totally agree that why I was posting the `kubernetes` way that seperate *health* probe and *readiness* probe. 

And even more have separate restart policy from *unhealthy* container and *failed result* during *readiness* period.

**Comment by vshcherb** [comment]
Readiness probe would be much better cause than you could report that service didn't start much quicker. Though it should be possible to specify max start time as well.

**Comment by krasi-georgiev** [comment]
readiness check  is the way to go - time based would allow false positives - 

For example a db container can't start for 3min this time , but next time it might take 10min if it does some updates , schema changes checks or even ithe db has grown significantly.

readiness check would be more flexilble.

**Comment by aaryn101** [comment]
I just want to chime in and state that a "readiness" feature would be extremely useful. I don't want to have to use Kubernetes for my production environments, but I will have to without this feature.

**Comment by dermatthes** [comment]
+1 for readyness. We're syncing databases/lbo which takes between 5 to 25 minutes (depending on load and bandwitdth). So a timed grace period is not enough for us.

**Comment by krasi-georgiev** [comment]
@thaJeztah  this can be closed in favour of #28938  right ?

**Comment by thaJeztah** [comment]
Yes, we can close this for now; we went for a grace period in the health-check, as it would resolve most use-cases, and doesn't add the complexity of a separate "readiness" option.

This keeps the road open for future additions (but no plans yet)
